Apps/NotSensorLogger2/app/src/main/java/com/example/notsensorlogger2/MainActivity.java

```diff
@@ -0,0 +1,353 @@
+                isDataCollectionActive = false;
+            }
+        });
+
+        Button btnOpenActivity2 = findViewById(R.id.open_activity2_button);
+        btnOpenActivity2.setOnClickListener(v -> {
+            Intent intent = new Intent(this, MainActivity2.class);
+            startActivity(intent);
+        });
+
+        sendDataRunnable = new Runnable() {
+            @Override
+            public void run() {
+                sendSensorData();
+                handler.postDelayed(this, SEND_INTERVAL);
+            }
+        };
+    }
+
+    /**
+     * Resumes the activity, checks necessary permissions, and resumes data collection if it was active before pausing.
+     */
+    @Override
+    protected void onResume() {
+        super.onResume();
+        checkAndRequestAudioPermission();
+        if (isDataCollectionActive) {
+            startDataCollection();
+        }
+    }
+
+    /**
+     * Pauses the activity and stops data collection if it was active.
+     */
+    @Override
+    protected void onPause() {
+        super.onPause();
+        if (isDataCollectionActive) {
+            stopDataCollection();
+        }
+    }
+
+    /**
+     * Starts the data collection from all active sensors and schedules periodic data sending to the server.
+     */
+    private void startDataCollection() {
+        messageId = 1;
+        for (SensorItem item : sensorItems) {
+            if (item.isToggled() && item.getHandler() != null) {
+                item.getHandler().start();
+            }
+        }
+        if (!isDataCollectionActive) {
+            handler.postDelayed(sendDataRunnable, SEND_INTERVAL);
+            isDataCollectionActive = true;
+        }
+    }
+
+    /**
+     * Stops the data collection from all sensors and cancels the periodic sending of data to the server.
+     */
+    private void stopDataCollection() {
+        for (SensorItem item : sensorItems) {
+            if (item.isToggled() && item.getHandler() != null) {
+                item.getHandler().stop();
+            }
+        }
+        if (isDataCollectionActive) {
+            handler.removeCallbacks(sendDataRunnable);
+            isDataCollectionActive = false;
+        }
+    }
+
+    /**
+     * Sends the collected sensor data to the server.
+     */
+    private void sendSensorData() {
+        JSONArray allSensorData = collectSensorData();
+        sendPostRequest(allSensorData);
+        clearAllSensorData();
+    }
+
+    /**
+     * Clears all sensor data from the memory.
+     */
+    private void clearAllSensorData() {
+        accelerometerHandler.clearSensorData();
+        gyroscopeHandler.clearSensorData();
+        magnetometerHandler.clearSensorData();
+        lightSensorHandler.clearSensorData();
+        humiditySensorHandler.clearSensorData();
+        pressureSensorHandler.clearSensorData();
+        gravitySensorHandler.clearSensorData();
+        locationTracker.clearSensorData();
+        if (soundLevelHandler != null) {
```

Comment:
Why the soundLevelHandler can be null? Is it initialized in onCreate? Same question for line 129 and line 291.

Reply:
even though if it has permission issue, the soundlevelhandler instance will still be created, just hte initAudioRecord() may not be run till the end. So almost impossible for soundLevelHandler to be null. If it is possible to be null, then the rest will have the risk to be null also